### ISSUE TITLE
Possible memory leak when inferencing BLOOM 176B

### ISSUE THREAD

**Issue by mayank31398** [huggingface/accelerate#614]
### System Info

```Shell
- `Accelerate` version: 0.11.0
- Platform: Linux-4.18.0-305.25.1.el8_4.x86_64-x86_64-with-glibc2.17
- Python version: 3.8.13
- Numpy version: 1.22.3
- PyTorch version (GPU?): 1.11.0a0+gitbc2c6ed (True)
- `Accelerate` default config:
	Not found
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

Script:
https://github.com/mayank31398/Megatron-DeepSpeed/blob/add-generation-server/scripts/inference/bloom-accelerate-server.py

Usage: python scripts/inference/bloom-accelerate-server.py --model_name bigscience/bloom --dtype bf16 --log_file data.log --host $ADDRESS --port $PORT

Memory blowup over time discussed here: https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/308#issuecomment-1205757494

### Expected behavior

```Shell
This memory leak should not occur I think.
```

**Comment by sgugger** [comment]
The model is used on our side on the inference API with the same options and without any memory leak, so I suspect the memory leak comes from somewhere else in the script and not Accelerate per se. I'm unfamiliar with Flask for instance, so not sure if it properly releases the memory of the objects no longer needed?

In any case, we'd need a smaller reproducer (ideally with a smaller model) to investigate more. We haven't seen anyone reporting any memory leak on large models as of now (including BLOOM, OPT, GPT-Neo-X or T5/T0pp)

**Comment by mayank31398** [comment]
Hmm, @sgugger can you tell which library are you guys using for the inference API, if not flask?

**Comment by sgugger** [comment]
We use `starlette` on our side.

**Comment by mayank31398** [comment]
If I call a torch.cuda.empty_cache() after this, then this happens:
<img width="560" alt="Screen Shot 2022-08-23 at 5 27 13 PM" src="https://user-images.githubusercontent.com/32954280/186152122-b5b16d85-1339-4560-9f94-774cb3254044.png">

**Comment by mayank31398** [comment]
Could it be that this is expected behaviour?
@ydshieh 
I am seeing a memory blowup with gpt2 also after replacing bigscience/bloom to gpt2

I am not sure if this is correct approach to measure memory in pytorch models

**Comment by mayank31398** [comment]
With pdb, I am seeing a blowup too.
But my guess would be this is not right way to measure memory since, I see something similar with GPT2 as well. I think the PyTorch memory allocator allocates some memory a priori for tensors and that is shown up in nvidia-smi.

**Comment by ydshieh** [comment]
> PyTorch memory allocator allocates some memory a priori for tensors and that is shown up in nvidia-smi.

This is correct. I also had an issue previously, see [here](https://github.com/huggingface/transformers/pull/16881#issuecomment-1106556203). In short, PyTorch won't always release GPU memory - it can re-use them later (faster operations), so it doesn't mean there is memory issue.

But after `empty_cache`, we should see the usage drops down, although might be partially.

**Comment by muellerzr** [comment]
You may also be able to get a bit more by doing garbage collection as well, after deleting the model in python 

E.g.:

```python
Import gc
del

**Comment by ydshieh** [comment]
@mayank31398 , when you say `a blowup too`, what it means?
We should measure the *difference* after the model inference against when the model is loaded to `cuda` but without running any input.

For GPT2, after loading the model, my GPU uses 1388 MB, and after inference, it goes to 1470 MB. I won't be surprised with this numbers however.

**Comment by ydshieh** [comment]
Hi, I have something for you @mayank31398 . Below is an example with `t5-large` (GPT2 is quite small to see the difference).

  - `model = model.to("cuda")`: 3662 MB
  - After the generation is done, but not return yet: 7746 MB
  - After the generation is done, and return to the main: 7746 MB
  - After empty cache: 3732M

So emptying cache can bring the GPU memory usage back to the point where model is loaded to GPU.

Here is the code
(The `FRANCE_ARTICLE`, `SHORTER_ARTICLE`, `IRAN_ARTICLE`, `ARTICLE_SUBWAY` are copied from https://github.com/huggingface/transformers/blob/0f257a87749e0a72bda260c6f319a45dae1e7c4d/tests/models/t5/test_modeling_t5.py#L924)
```python
import pdb

from transformers import AutoModelForSeq2SeqLM, AutoTokenizer
import torch


def run(model, dct):

    hypotheses_batch = model.generate(
        **dct,
        num_beams=4,
        length_penalty=2.0,
        max_length=142,
        min_length=56,
        no_repeat_ngram_size=3,
        do_sample=False,
        early_stopping=True,
    )

    print("gen. done")
    pdb.set_trace()


if __name__ == "__main__":

    ckpt = "t5-large"
    tokenizer = AutoTokenizer.from_pretrained(ckpt)

    model = AutoModelForSeq2SeqLM.from_pretrained(ckpt)
    model.config.update(model.config.task_specific_params["summarization"])

    dct = tokenizer(
        [model.config.prefix + x for x in [FRANCE_ARTICLE, SHORTER_ARTICLE, IRAN_ARTICLE, ARTICLE_SUBWAY]],
        padding="max_length",
        truncation=True,
        return_tensors="pt",
    )
    print("input / model on cpu")
    pdb.set_trace()

    dct = dct.to("cuda")
    print("input to gpu")
    pdb.set_trace()

    model = model.to("cuda")
    print("model to gpu")
    pdb.set_trace()

    run(model, input)
    print("model run done")
    pdb.set_trace()

    torch.cuda.empty_cache()
    print("clear done")
    pdb.set_trace()
```

**Comment by mayank31398** [comment]
> 

Deleting the model is not an option for me.
I am trying to use the model in a server setting for a lot of folks.
Related PR: https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/328

**Comment by mayank31398** [comment]
> You may also be able to get a bit more by doing garbage collection as well, after deleting the model in python
> 
> E.g.:
> 
> ```python
> import gc
> del model
> gc.collect()
> ```
> 
> (Also sorry for accidently closing, on mobile and hit the wrong button!)

Deleting the model is not an option for me.
I am trying to use the model in a server setting for a lot of folks.
Related PR: https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/328

**Comment by mayank31398** [comment]
Yes, thanks I think Ill try to see the memory usage over time by running in a for loop or something.
To see how this changes memory (both in server and non-server settings).

**Comment by stas00** [comment]
what @ydshieh said and more: 

to track **real** memory usage / debug potential leaks always:
1. call `gc.collect()` first - since python's GC is scheduled and w/o it you might miss object and its associated memory release
2. then clear the cuda cache 
3. measure

but don't do any of the above for production.

**Comment by mayank31398** [comment]
This is not an issue anymore. Thanks for helping guys.
Closing this :)

**Comment by ydshieh** [comment]
@mayank31398 Would you like to share what works for you :-) 🙏

**Comment by mayank31398** [comment]
I converted my server to flask and ran with gunicorn with 1 worker.
This serializes all requests however
